File: backend/src/github/github.service.ts
import { HttpException, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

interface GithubRepo {
  name: string;
  html_url: string;
  description: string | null;
  language: string | null;
  stargazers_count: number;
  forks_count: number;
  fork: boolean;
  updated_at: string;
}

@Injectable()
export class GithubService {
  private readonly baseUrl = 'https://api.github.com';

  constructor(private readonly config: ConfigService) {}

  private authHeaders(): Record<string, string> {
    const token = this.config.get<string>('GITHUB_TOKEN');
    const headers: Record<string, string> = {
      Accept: 'application/vnd.github+json',
      'User-Agent': 'DevAgent',
    };
    if (token) headers.Authorization = `Bearer ${token}`;
    return headers;
  }

  private async request<T>(path: string): Promise<T> {
    const res = await fetch(`${this.baseUrl}${path}`, { headers: this.authHeaders() });
    if (!res.ok) {
      const body = await res.text();
      throw new HttpException(
        `GitHub API request failed (${res.status}): ${body}`,
        res.status >= 400 && res.status < 500 ? res.status : 502,
      );
    }
    return res.json() as Promise<T>;
  }

  async getProfile(username: string) {
    return this.request(`/users/${encodeURIComponent(username)}`);
  }

  async getRepos(username: string, limit = 100) {
    const repos = await this.request<GithubRepo[]>(
      `/users/${encodeURIComponent(username)}/repos?per_page=${limit}&sort=updated`,
    );
    return repos.map((r) => ({
      name: r.name,
      url: r.html_url,
      description: r.description,
      language: r.language,
      stars: r.stargazers_count,
      forks: r.forks_count,
      fork: r.fork,
      updatedAt: r.updated_at,
    }));
  }

  /** Combined profile + language breakdown + top repos, used by the portfolio generator too. */
  async getSummary(username: string) {
    const [profile, repos] = await Promise.all([
      this.getProfile(username),
      this.getRepos(username),
    ]);

    const languageCounts: Record<string, number> = {};
    for (const r of repos) {
      if (r.language) languageCounts[r.language] = (languageCounts[r.language] || 0) + 1;
    }

    const topRepos = [...repos]
      .filter((r) => !r.fork)
      .sort((a, b) => b.stars - a.stars)
      .slice(0, 6);

    return {
      profile,
      totalRepos: repos.length,
      languages: Object.entries(languageCounts)
        .sort((a, b) => b[1] - a[1])
        .map(([language, count]) => ({ language, count })),
      topRepos,
    };
  }
}

File: backend/src/main/java/com/devagent/github/GithubService.java
package com.devagent.github;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class GithubService {

    private static final String BASE = "https://api.github.com";
    private final OkHttpClient httpClient = new OkHttpClient();
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Value("${github.token:}")
    private String githubToken;

    private Request.Builder builder(String url) {
        Request.Builder b = new Request.Builder()
                .url(url)
                .header("Accept", "application/vnd.github+json")
                .header("User-Agent", "DevAgent");
        if (githubToken != null && !githubToken.isBlank()) {
            b.header("Authorization", "Bearer " + githubToken);
        }
        return b;
    }

    private <T> T fetch(String url, TypeReference<T> type) {
        try (Response res = httpClient.newCall(builder(url).build()).execute()) {
            if (!res.isSuccessful()) {
                throw new ResponseStatusException(HttpStatus.valueOf(res.code()),
                        "GitHub API error: " + res.code());
            }
            return objectMapper.readValue(res.body().string(), type);
        } catch (ResponseStatusException e) {
            throw e;
        } catch (IOException e) {
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY,
                    "Failed to reach GitHub API: " + e.getMessage());
        }
    }

    public Map<String, Object> getProfile(String username) {
        return fetch(BASE + "/users/" + username,
                new TypeReference<>() {});
    }

    public List<Map<String, Object>> getRepos(String username, int limit) {
        return fetch(BASE + "/users/" + username + "/repos?per_page=" + limit + "&sort=updated",
                new TypeReference<>() {});
    }

    public Map<String, Object> getSummary(String username) {
        Map<String, Object> profile = getProfile(username);
        List<Map<String, Object>> rawRepos = getRepos(username, 100);

        // Language breakdown
        Map<String, Long> langCounts = rawRepos.stream()
                .filter(r -> r.get("language") != null)
                .collect(Collectors.groupingBy(r -> (String) r.get("language"),
                        Collectors.counting()));

        List<Map<String, Object>> languages = langCounts.entrySet().stream()
                .sorted(Map.Entry.<String, Long>comparingByValue().reversed())
                .map(e -> Map.<String, Object>of("language", e.getKey(), "count", e.getValue()))
                .toList();

        // Top repos — non-forks by stars
        List<Map<String, Object>> topRepos = rawRepos.stream()
                .filter(r -> !Boolean.TRUE.equals(r.get("fork")))
                .sorted(Comparator.comparingInt(
                        (Map<String, Object> r) -> (Integer) r.getOrDefault("stargazers_count", 0))
                        .reversed())
                .limit(6)
                .map(r -> {
                    Map<String, Object> repo = new LinkedHashMap<>();
                    repo.put("name", r.get("name"));
                    repo.put("url", r.get("html_url"));
                    repo.put("description", r.get("description"));
                    repo.put("language", r.get("language"));
                    repo.put("stars", r.getOrDefault("stargazers_count", 0));
                    repo.put("forks", r.getOrDefault("forks_count", 0));
                    repo.put("fork", r.getOrDefault("fork", false));
                    repo.put("updatedAt", r.get("updated_at"));
                    return repo;
                })
                .toList();

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("profile", profile);
        result.put("totalRepos", rawRepos.size());
        result.put("languages", languages);
        result.put("topRepos", topRepos);
        return result;
    }
}
